### PR TITLE
Possibility to use their own metrics

### PR DESCRIPTION
**Pull request by CeVague**

### CODE REVIEW

File: wizard/management/commands/load_default_metrics.py [hunk 1]
```diff
@@ -7,56 +7,97 @@
...
-                   },
-    'auc_multilabel': {'description': """Normalized Area under ROC curve for multi-labeled data""",
-                       'is_default': True,
-                       'classification': True,
-                       },
-    'pac_binary': {'description': """Probabilistic Accuracy based on log_loss for binary data""",
-                   'is_default': True,
-                   'classification': True,
-                   },
+    'example': {
```
Comment (ckcollab) [inline wizard/management/commands/load_default_metrics.py]: Fix formatting/indentation here?

File: wizard/models.py [hunk 2]
```diff
@@ -247,7 +247,8 @@ def clean(self):
         cols = columns_count_first_line(self.raw_content)
 
         if not columns_all_the_same_count(self.raw_content, cols):
-            raise InvalidAutomlFormatException("Number of cols non coherent in %s" % self.raw_content)
+            pass
+            #raise InvalidAutomlFormatException("Number of cols non coherent in %s" % self.raw_content)
```
Comment (ckcollab) [inline wizard/models.py]: pull this?

File: wizard/resources/__init__.py [hunk 3]
```diff
@@ -3,9 +3,24 @@
...
 
-def build_default_scoring(name, dest):
+def build_default_scoring(metric, dest):
     p = fs.here(__file__, 'default_scoring')
     fs.copy_dir_content(p, dest)
 
+    path_new = os.path.join(dest, 'libscores.py')
+    path_old = os.path.join(dest, 'libscores_tmp.py')
+
+    with open(path_new,'w') as new_file:
```
Comment (ckcollab) [inline wizard/resources/__init__.py]: space after comma (we can maybe add a linter to this repository?)

File: wizard/resources/__init__.py [hunk 4]
```diff
@@ -3,9 +3,24 @@
...
+                    new_file.write(metric.code)
+                else:
+                    new_file.write(line)
+
+    #Remove original file
+    os.remove(path_old)
+
     with open(os.path.join(dest, 'metric.txt'), 'w') as f:
-        print(name, file=f)
+        print(metric.name, file=f)
```
Comment (ckcollab) [inline wizard/resources/__init__.py]: is this debugging code?

File: wizard/templates/wizard/data/editor.html [hunk 5]
```diff
@@ -39,8 +39,8 @@
             </p>
 
             <p>
-                Only <code>automl, csv and arff</code> archives are currently supported. They are defined as a zip
-                file containing at least:
+                Only <code>AutoML, CSV and ARFF</code> archives are currently supported. All archives will be converted
```
Comment (ckcollab) [inline wizard/templates/wizard/data/editor.html]: Should probably be:

`<code>AutoML</code>, <code>CSV</code> and <code>ARFF</code>`

File: wizard/templates/wizard/metric/editor.html [hunk 6]
```diff
@@ -1,29 +1,115 @@
...
+                            <script>
+                                var metrics = {
+                                {% for metric in public_metrics %}
+                                    {{ metric.pk }} : ["{{ metric.name }}", "{{ metric.description }}", "{{ metric.code|escapejs }}"],
+                                {% endfor %}
+                                }
+
+                                function fillWithMetric(s) {
+                                    metric = metrics[s.value];
+                                    document.getElementById('name').value = metric[0];
```
Comment (ckcollab) [inline wizard/templates/wizard/metric/editor.html]: nice

File: wizard/views.py [hunk 7]
```diff
@@ -316,95 +317,59 @@ def data_picker(request, pk, cant_delete = False):
...
         if cant_delete:
-            from django.contrib import messages
             messages.error(request, "This dataset can't be deleted : another challenge use it.")
 
         return render(request, 'wizard/data/picker.html', context=context)
 
 
-def metric_picker(request, pk):
+def metric(request, pk):
+
```
Comment (ckcollab) [inline wizard/views.py]: remove extra newline?

File: wizard/views.py [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 from django.contrib.auth.decorators import login_required
 from django.contrib.auth.mixins import LoginRequiredMixin
+from django.contrib import messages
```
Comment (ckcollab) [inline wizard/views.py]: does this get used?

File: wizard/views.py [hunk 9]
```diff
@@ -316,95 +317,59 @@ def data_picker(request, pk, cant_delete = False):
...
-    current_flow = flow.MetricFlowItem
+        # If it's here first metric or it's a default one, we create a new one
+        if (not c.metric is None) and (not c.metric.is_default):
+            new_metric = get_object_or_404(MetricModel, id=c.metric.id)
+        else:
+            new_metric.owner = request.user
 
-    def get_form(self, form_class=None):
-        form = super().get_form(form_class=form_class)
+        new_metric.name = request.POST['name']
```
Comment (ckcollab) [inline wizard/views.py]: Should use a Django Model Form! :)

File: wizard/views.py (hunk 8 above)
Comment (CeVague) [inline wizard/views.py]: Yes, two times

File: wizard/resources/__init__.py (hunk 4 above)
Comment (CeVague) [inline wizard/resources/__init__.py]: No, it output on the file ("metric.txt")

File: wizard/models.py (hunk 2 above)
Comment (CeVague) [inline wizard/models.py]: Yeah, one the the default dataset don't pass this test.
So I don't know if it's the dataset witch have a problem or if this test is useless (event if it was on a TODO comment)

File: wizard/management/commands/load_default_metrics.py (hunk 1 above)
Comment (CeVague) [inline wizard/management/commands/load_default_metrics.py]: I add the "code" part and a default metric

File: wizard/templates/wizard/data/editor.html (hunk 5 above)
Comment (CeVague) [inline wizard/templates/wizard/data/editor.html]: Yeah, it's better like you say

File: wizard/resources/__init__.py (hunk 4 above)
Comment (ckcollab) [inline wizard/resources/__init__.py]: Can we use `f.write` or something instead? more explicit :)

File: wizard/views.py (hunk 8 above)
Comment (ckcollab) [inline wizard/views.py]: 👍  in the future we should probably avoid that module (not very popular), but if it's already used here that's fine.

File: wizard/models.py (hunk 2 above)
Comment (ckcollab) [inline wizard/models.py]: > Yeah, one the the default dataset don't pass this test.

Ah, okay :(

File: wizard/resources/__init__.py (hunk 4 above)
Comment (CeVague) [inline wizard/resources/__init__.py]: Yeah, good idea ^^

**Comment by ckcollab** [comment]
Whoa! Thanks for fixing all of the pep8 problems!!! 

Last 2 things for this PR, I believe, are:

- [ ] Set up http://circleci.com for this project so tests run automatically
- [ ] Add a test to `circle.yml` to run `flake8`, like below..

```yaml
test:
  override:
    - flake8 .  
    - python manage.py test
```

This will have `flake8` run before regular django tests on Circle

**Comment by ckcollab** [comment]
Hmm, can you get admin access to this repo, so we can move it and such? I'm not sure we can integrate CircleCI without an admin...

File: wizard/migrations/0079_challengemodel_build_at.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('wizard', '0078_datasetmodel_updated_at'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='challengemodel',
+            name='build_at',
```
Comment (ckcollab) [inline wizard/migrations/0079_challengemodel_build_at.py]: Be careful to not make too many migrations, can get unwieldy after too long :) Try to think out the schema ahead of time, although that's often very difficult/impossible, so we have as few migrations as possible.

**Pull request merged**